### PR TITLE
Add Mission Control: a live grid of every agent pane

### PR DESCRIPTION
**Pull request by parsakhaz**
## What you are approving

**Mission Control**: a new top-level view with one tile per agent pane, so several
agents can be watched at once. **@0x92 wrote the feature** in
[#382](https://github.com/dcouple/Pane/pull/382); the commits after theirs are a
rename and several rounds of review fixes, kept separate so their work stays
legible underneath.

- **Where to look hardest:** one change reaches outside this feature into shared
  terminal flow control, so it affects every terminal, not just the new grid. It
  is the first item under "Defects found and fixed" below, "Flow control credited
  the same bytes twice".
- **Risk that already surfaced once:** that same change introduced a jam that
  dropped every terminal panel's ack. A later review caught it, it is fixed, and a
  test now fails without the fix. Worth a second pair of eyes precisely because it
  is the one thing here that can affect someone who never opens Mission Control.
- **Everything else** stays inside the new view.
- **Verified by driving the real app** with live agents, not only by tests:
  promotion, focus, collapse, grouping, density, stopped tiles, and widths from
  380px to 1440px.

If you have ten minutes: watch the clip, read the flow-control defect, then read
the five known gaps at the end.

## Description

Mission Control is a top-level view that shows every agent pane at once, so you
can watch several agents run and see which one is waiting without visiting each
session in turn. Three people have asked for this.

The base is **[PR #382](https://github.com/dcouple/Pane/pull/382) by @0x92**,
cherry-picked so the commit and its authorship stay intact. Everything after it is
a separate commit on top. The feature ships as **Mission Control**, the name we
use for this surface across the product, replacing the original "Agent Fleet".

<a href="https://raw.githubusercontent.com/dcouple/Pane/mission-control-evidence/clips/mission-control.mp4">Mission Control with several agents running</a>

<sub>Several Claude Code agents working at once in a scratch repository. The clip
regroups by status and back to project, promotes a tile to a live terminal under
the pointer, and switches column counts.
[Full-resolution MP4](https://raw.githubusercontent.com/dcouple/Pane/mission-control-evidence/clips/mission-control.mp4)
· [full-resolution still](https://raw.githubusercontent.com/dcouple/Pane/mission-control-evidence/clips/mission-control.png)
· sources on [`mission-control-evidence`](https://github.com/dcouple/Pane/tree/mission-control-evidence).</sub>

## What it does

Open it from the sidebar rail or from the expanded tree, and you get:

- A grid of tiles, one per agent pane, with a status accent, branch and age,
  grouped by project, status or agent.
- Snapshots that are cheap ANSI-stripped text by default. A tile becomes a real
  xterm under the pointer, or all of them with "All live".
- Typing into a focused tile without leaving the grid, optionally expanding it to
  the agent's full screen.
- Folding projects away, jumping to the next blocked agent with `N`, moving with
  the arrow keys, closing a pane.

Two constraints from the original PR are load-bearing, and both are documented
where they are enforced:

- **A tile renders at exactly the PTY's dimensions.** Agent TUIs paint with
  absolute cursor positioning, and a mismatched width wraps every line. A tile
  therefore never resizes the agent's PTY, which is why it fits by choosing a font
  size instead.
- **A tile stays silent on the agent's terminal queries**, leaving the real panel
  as the single voice on that PTY.

## How it got here

Ten commits, oldest first below. The first is the contribution; the rest are the
cleanup, split so each is reviewable on its own. Skip this table unless you want
the trail.

| Commit | What |
| --- | --- |
| `Add Agent Fleet: a live grid of every agent pane` (@0x92) | The feature, cherry-picked from #382 with conflicts resolved against current main. |
| `Rename Agent Fleet to Mission Control` | Strings, sidebar entries, `activeView`, the two IPC channels, the CSS hook, files and symbols. |
| `Refactor … to current main's gates and fix the viewer sweep` | Lint conformance, boundary-decoded payloads, guarded persisted options, one real defect. |
| `Simplify Mission Control` | Shared derivations, de-duplicated markup, removed payload fields nothing reads. |
| `Fix the stopped-tile transcript, the terminal font, and idle cost` | Six review findings. |
| `Address the review: one ack consumer, and four correctness fixes` | Codex review. |
| `Fix a flow-control jam and six findings from a fresh review` | Including a jam this branch introduced. |
| `Fit tile content to the tile, snapshot and live alike` | The overflow fix. |
| `Scale the column cap and the expanded tile to the display` | Responsiveness. |
| `Simplify pass: stop bypassing the shared fit, and drop what nothing reads` | Four-angle quality pass. |

### The conflict resolution

Main moved under the branch: #472 (title-bar context), #476 (Window Controls
Overlay), #478 and #480 (animations) all touched the same shell chrome. The one
substantive resolution is in `main/src/ipc/runpane.ts`: #382 extracts three
screen-text helpers into `services/panels/terminalScreenText.ts`, and main had
independently refactored those same helpers onto the boundary decoder in #420.
The extracted module carries main's version.

### Defects found and fixed along the way

Every one of these was caught by review before merge rather than by a user after
it. They are listed in full because the reasons are worth knowing, and all of
them are closed.

- **Flow control credited the same bytes twice.** This is the one that reaches
  outside the feature. `pendingBytes` is one counter per PTY, and every viewer
  that acks a broadcast chunk credits it again, so the high watermark was never
  reached and the PTY was never paused. VS Code's terminal takes the same shape
  (one counter, a single designated consumer) and never puts two xterms on one
  PTY. Acks now carry a viewer id and main designates one consumer per panel.
  This also closes the same double-ack between a panel and the Remote PWA, which
  predates this feature.

  The first attempt at that fix jammed every terminal, which is the risk named
  at the top. `terminal:setVisibility` scoped a bare viewer id to `local:<id>`
  and the new `terminal:ack` passed the same id through raw, so a panel
  registered under one string and acknowledged under another. Every ack from a
  visible terminal was discarded, `pendingBytes` climbed to the high watermark,
  and the safety timer force-resumed without clearing the counter, so a busy
  build would have delivered roughly one chunk every five seconds. Scoping now
  happens in one place, and a test registers scoped, acknowledges bare, and
  fails without the fix. It is worth knowing that the first version of that test
  passed either way, because it fed already-scoped ids to both sides.
- **The stale-viewer sweep never matched anything.** Main swept the prefix
  `mission-control:` while the renderer minted `missionControl:<uuid>`, and the
  matcher appends its own separator, so a prefix ending in a colon can never
  match. The prune timer was inert and a crashed renderer left every hovered
  panel pinned visible. Both processes read one shared constant now.
- **A stopped agent's tile showed a mangled transcript.**
  `alternateScreenBuffer` is the raw PTY byte stream, and a full-screen TUI paints
  it with absolute cursor positioning, so stripping the escapes ran every word
  together. Panels persist the emulator's laid-out `screenText` now, exit persists
  before teardown, and `runpane panels screen` gets the same fix.
- **Closing Pane Chat killed it and reported success.** `panels:delete` destroyed
  the terminal, then declined the permanent panel and returned quietly. It refuses
  before teardown now, and the grid offers no close action for a permanent panel.
- Four smaller ones, all fixed: focusing a tile rebuilt and re-hydrated its
  terminal; a finished agent kept a live tile; deleting an agent in a background
  session wedged the sidebar badge; past the 64-panel cap a tile could go live
  with no PTY dimensions and fall back to 80x24.

### Responsiveness

The expanded tile cut its terminal off mid-word, because the font was fitted to
the whole wrapper while the rendered terminal also carries about 26px of its own
chrome. Snapshot bodies never fitted at all: they rendered at a fixed 11px and let
the tile clip whatever missed, so an unpromoted tile clipped silently at every
width while a live one scaled. Both paths share one fit module now, and the
clipped-right fade is shown by both rather than looking like a rendering fault in
either.

The column cap is derived from what a conventional 80-column screen needs at the
legibility floor, measured against the configured terminal font, so the same
density setting means more columns on a large display and fewer on a laptop. The
4x setting asks for four columns and the grid gives as many as fit:

| Viewport width (px) | Columns at the 4x setting | Tile width |
| --- | --- | --- |
| 3840 | 4 | 927px |
| 2560 | 4 | 607px |
| 1920 | 4 | 447px |
| 1440 | 2 | 664px |
| 800 | 1 | 698px |

Measured from 380px to 1440px: zero elements escaping the window, zero horizontal
document scroll, in both the normal grid and the expanded state.

### Idle cost

Idle cost follows the existing `terminalPowerMode` setting rather than a policy of
this view's own. Performance keeps the poll and the hovered tile alive while the
window is behind something else; Battery Saver drops them, which is what
`TerminalPanel` already does with the same setting.

## Type of Change
- [x] New feature (non-breaking change which adds functionality)

## Checklist
- [x] I have read the [CONTRIBUTING.md](../CONTRIBUTING.md) guidelines
- [x] My code follows the code style of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (AGENTS.md)
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have run `pnpm typecheck` and `pnpm lint` locally
- [x] I have tested the Electron app locally

## Critical Areas Modified
- [x] Session output handling (requires explicit permission). Flow-control
      acknowledgement of PTY output is now gated on a designated viewer per
      panel, and a panel persists the emulator's screen text when its PTY exits.
      Both reach every terminal, not only Mission Control, so this box is
      checked deliberately rather than by omission.
- [x] State management/IPC events (two new `mission-control:` channels, registered
      through the shared command registry and covered by `daemonRegistryBindings.test.ts`)

## Manual tests

Driven against a development build on an isolated `PANE_DIR`, with a scratch
repository and several Claude Code agents working at once.

| What | Result |
| --- | --- |
| Open Mission Control from the sidebar rail and from the expanded tree | Both reach the grid |
| Several live agents plus Pane Chat | One tile each, header counts agents and how many are working |
| Tile content while agents work | Snapshots advance with the agents' output |
| Hover a tile | Exactly one tile promotes to a live xterm, built once |
| Click a tile, then Collapse | Expands to the agent's screen and returns to grid size at 1440, 1100 and 800 |
| An agent asking a question | Tile sorts to the front with the "Needs input" accent |
| Group by status, agent, none, and back to project | Regroups in place, counts follow |
| Column counts 1x through 4x | Honoured as far as the width allows, then capped |
| Window widths 380px to 1440px | No overflow in either the grid or the expanded state |
| Exit an agent, then read its tile | Shows the agent's real last screen, correctly laid out |
| Pane Chat tile | No close action, and opening it goes to Pane Chat |
| `pnpm lint`, `pnpm typecheck` | Pass |
| `pnpm --filter main exec vitest run` | 68 files, 644 tests pass |
| `pnpm --filter frontend exec vitest run` | 28 files, 290 tests pass |
| `pnpm test:ci:minimal` | 27 Playwright tests pass |

### AGENTS.md picks up an Architecture Invariants section

It arrives from #382 and reaches well past this feature: the 7-file IPC dance, the
`claude` shim that `pnpm install` writes, the Windows dev data directory,
`CommandRunner`, where schema actually lives, panel types, the shared WebGL texture
atlas, and the lint rules that fail a PR. It is worth reading rather than skimming.

### Known gaps, deliberately left for later

- Tile status accents use their own colour map rather than `StatusAccentBar`, so a
  working agent's tile accent differs from the dot beside it. The fork looks
  deliberate (a grid of green idle tiles would be noise) but it should be settled
  one way or the other.
- `MissionControlAgentType` covers `claude` and `codex`, so Cursor panes list as
  "Agent" under "Other agents" and cannot be grouped by type.
- The grid's `N` shortcut is a raw document listener rather than a registered
  hotkey, so it is invisible to the shortcuts overlay and ignores the
  keyboard-shortcuts setting.
- The agent roster query filters on `json_extract` over every terminal panel's
  state blob, which is where scrollback lives. It wants an indexed column.
- Preload duplicates the daemon-owned channel lists; that is
  [#485](https://github.com/dcouple/Pane/issues/485).

## Latest automated QA

**Passed at `59189e2c`.** The maintained minimal Electron suite passed 27/27 tests. A focused mocked Mission Control journey also passed for a populated grid, status grouping, four-column selection, group collapse, and a 420px viewport with no document-level horizontal overflow. Evidence uses synthetic repository and agent data only.

Remaining human check: real PTY interaction with live agents, especially terminal-query replies and hover/focus promotion. The PR's existing manual test record covers those paths on the earlier head; this run did not repeat them against a real agent process.

### CODE REVIEW

File: frontend/src/hooks/useMissionControlTerminal.ts [hunk 1]
```diff
@@ -0,0 +1,561 @@
...
+      if (!data) return;
+      terminal.write(data.output, () => {
+        if (disposed) return;
+        // Follow new output, but never yank the view away from someone who
+        // scrolled up to read.
+        if (!pinnedByUser) scrollToLastContent(terminal);
+        syncOverflow();
+      });
+      // Flow control is byte-counted per panel; a viewer that writes must ack.
+      void window.electronAPI.invoke('terminal:ack', panelId, terminalOutputByteLength(data.output)).catch(() => {});
```
[the remark this answers is not in the document]
Comment (parsakhaz) [reply frontend/src/hooks/useMissionControlTerminal.ts]: Fixed in 83ceb1f. Acks now carry a viewer id and the main process designates a single consumer per panel from the viewers it already tracks: a real panel or remote runtime when one is watching, a preview tile only when it is the only viewer (otherwise nobody would ack and the PTY would stay paused). Callers that send no viewer id are still credited, since every ack path predates them. This closes the same double-ack between a terminal panel and the Remote PWA, which predates this feature. Tests in terminalPanelManager.test.ts cover all three cases, and the double-ack one fails without the gate.

File: frontend/src/components/missionControl/MissionControlView.tsx [hunk 2]
```diff
@@ -0,0 +1,813 @@
...
+
+  // --- Closing a pane ---
+
+  const handleConfirmClose = useCallback(async () => {
+    const tile = pendingClose;
+    if (!tile) return;
+
+    setClosing(true);
+    try {
+      await panelApi.deletePanel(tile.panelId);
```
[the remark this answers is not in the document]
Comment (parsakhaz) [reply frontend/src/components/missionControl/MissionControlView.tsx:798]: Fixed in 83ceb1f, at the source rather than in the caller: panels:delete now refuses a permanent panel before any teardown, instead of destroying the terminal and then reporting success for a delete that deletePanel declined. Mission Control also stops rendering a close action for a permanent panel, so Pane Chat's tile has none.

File: frontend/src/components/missionControl/MissionControlView.tsx [hunk 3]
```diff
@@ -0,0 +1,813 @@
...
+    }
+  }, []);
+
+  // Agents appearing or disappearing shows up as agent-status churn; refresh
+  // the roster on it rather than polling the (heavier) list endpoint. This also
+  // covers the initial load, since the effect runs on mount.
+  const agentStatusKeys = useMemo(() => Object.keys(agentStatus).sort().join(','), [agentStatus]);
+  useEffect(() => {
+    void loadAgents();
+  }, [agentStatusKeys, loadAgents]);
```
[the remark this answers is not in the document]
Comment (parsakhaz) [reply frontend/src/components/missionControl/MissionControlView.tsx]: Fixed in 83ceb1f. The roster now keys on panel id and state rather than the id set alone, so an agent that exits (same id, new state) reloads isLive instead of leaving its tile live and interactive.

File: frontend/src/components/missionControl/MissionControlView.tsx [hunk 4]
```diff
@@ -0,0 +1,813 @@
...
+      }));
+  }, [groups, focusedPanelId]);
+
+  // A folded-away group costs nothing: its panels leave the snapshot batch, so
+  // the emulators behind them are never woken for a tile nobody can see.
+  const panelIds = useMemo(
+    () => groups
+      .filter(group => !collapsedGroups.has(group.key))
+      .flatMap(group => group.tiles.map(tile => tile.panelId))
+      .slice(0, MAX_MISSION_CONTROL_SNAPSHOT_PANELS),
```
[the remark this answers is not in the document]
Comment (parsakhaz) [reply frontend/src/components/missionControl/MissionControlView.tsx]: Fixed in 83ceb1f. A tile only goes live when its PTY dimensions are known, which past the cap they are not. That matches the footer note already telling the user live previews stop at the first 64 agents, and it avoids replaying a differently sized TUI into a guessed 80x24 grid.

File: frontend/src/components/missionControl/MissionControlView.tsx [hunk 5]
```diff
@@ -0,0 +1,813 @@
...
+    if (!tile) return;
+
+    setClosing(true);
+    try {
+      await panelApi.deletePanel(tile.panelId);
+      if (focusedPanelId === tile.panelId) handleCloseFocus();
+      setLiveTileId(current => (current === tile.panelId ? null : current));
+      setSelectedPanelId(current => (current === tile.panelId ? null : current));
+      setPendingClose(null);
+      await loadAgents();
```
[the remark this answers is not in the document]
Comment (parsakhaz) [reply frontend/src/components/missionControl/MissionControlView.tsx:809]: Fixed in 83ceb1f. The view clears the panel's agent status as part of the delete it just performed, so a blocked agent in a background session no longer leaves the sidebar badge counting it.

File: frontend/src/components/missionControl/MissionControlView.tsx [hunk 6]
```diff
@@ -0,0 +1,866 @@
...
+/**
+ * Live grid of every agent pane across every session and project.
+ *
+ * Tiles poll a batched, ANSI-stripped snapshot while this view is visible, and
+ * exactly one tile at a time is promoted to a real read-only terminal on hover
+ * or focus. That cap is deliberate: xterm instances sharing a font and theme
+ * share a WebGL texture atlas, and a grid of live terminals would exhaust
+ * Chromium's GL contexts (see `TerminalPanel`'s header comment).
+ */
+export function MissionControlView() {
```
Comment (parsakhaz) [inline frontend/src/components/missionControl/MissionControlView.tsx:216]: Not changed. The component is long, and two pieces (the segmented control and the status pill) are already extracted. Splitting the rest means a header component with a dozen props threaded through it, which trades one readability problem for another, so this stays as an accepted warning rather than a refactor done to satisfy a line count.

File: frontend/src/components/missionControl/MissionControlView.tsx [hunk 7]
```diff
@@ -0,0 +1,866 @@
...
+        tiles: [...group.tiles].sort(
+          (a, b) => rankOf(tileRank, a.panelId) - rankOf(tileRank, b.panelId)
+        ),
+      }));
+  }, [groups, focusedPanelId]);
+
+  // A folded-away group costs nothing: its panels leave the snapshot batch, so
+  // the emulators behind them are never woken for a tile nobody can see.
+  const panelIds = useMemo(
+    () => groups
```
Comment (parsakhaz) [inline frontend/src/components/missionControl/MissionControlView.tsx]: Fixed in 83ceb1f: this builds its list in one pass now.

File: frontend/src/components/missionControl/MissionControlView.tsx [hunk 8]
```diff
@@ -0,0 +1,866 @@
...
+      .filter(group => !collapsedGroups.has(group.key))
+      .flatMap(group => group.tiles.map(tile => tile.panelId))
+      .slice(0, MAX_MISSION_CONTROL_SNAPSHOT_PANELS),
+    [groups, collapsedGroups]
+  );
+  const panelIdsKey = panelIds.join(',');
+
+  // Poll snapshots only while this view is actually on screen. A background
+  // poll would keep every agent's emulator warm for nothing.
+  useEffect(() => {
```
Comment (parsakhaz) [inline frontend/src/components/missionControl/MissionControlView.tsx:475]: Not changed, and I believe this is a false positive. The setter is already behind the effect's cancellation flag (`if (cancelled || ...) return` immediately after the await, with `cancelled` set in the cleanup), which is exactly the guard the rule asks for.

File: frontend/src/components/missionControl/MissionControlView.tsx [hunk 9]
```diff
@@ -0,0 +1,866 @@
...
+    setActivePanelInStore(tile.sessionId, tile.panelId);
+    await setActiveSession(tile.sessionId);
+    navigateToSessions();
+  }, [setActiveSession, setActivePanelInStore, navigateToSessions]);
+
+  // --- Keyboard navigation ---
+
+  /** Every tile on screen, in reading order, ignoring folded groups. */
+  const visibleTiles = useMemo(
+    () => displayedGroups
```
Comment (parsakhaz) [inline frontend/src/components/missionControl/MissionControlView.tsx]: Fixed in 83ceb1f: this builds its list in one pass now.

File: frontend/src/components/missionControl/MissionControlView.tsx [hunk 10]
```diff
@@ -0,0 +1,890 @@
...
+/**
+ * Live grid of every agent pane across every session and project.
+ *
+ * Tiles poll a batched, ANSI-stripped snapshot while this view is visible, and
+ * exactly one tile at a time is promoted to a real read-only terminal on hover
+ * or focus. That cap is deliberate: xterm instances sharing a font and theme
+ * share a WebGL texture atlas, and a grid of live terminals would exhaust
+ * Chromium's GL contexts (see `TerminalPanel`'s header comment).
+ */
+export function MissionControlView() {
```
Comment (parsakhaz) [inline frontend/src/components/missionControl/MissionControlView.tsx:216]: Same answer as the earlier thread: accepted as a warning. Two pieces are already extracted; splitting the rest means a header component with a dozen props, which is not an improvement.

File: frontend/src/components/missionControl/MissionControlView.tsx [hunk 11]
```diff
@@ -0,0 +1,890 @@
...
+        ids.push(tile.panelId);
+      }
+    }
+    return ids;
+  }, [groups, collapsedGroups]);
+  const panelIdsKey = panelIds.join(',');
+
+  // Poll snapshots only while this view is actually on screen. A background
+  // poll would keep every agent's emulator warm for nothing.
+  useEffect(() => {
```
Comment (parsakhaz) [inline frontend/src/components/missionControl/MissionControlView.tsx:475]: Same answer as the earlier thread: this setter is already behind the effect's cancellation flag, which is the guard the rule asks for.

File: frontend/src/components/missionControl/MissionControlView.tsx [hunk 12]
```diff
@@ -0,0 +1,981 @@
...
+/**
+ * Live grid of every agent pane across every session and project.
+ *
+ * Tiles poll a batched, ANSI-stripped snapshot while this view is visible, and
+ * exactly one tile at a time is promoted to a real read-only terminal on hover
+ * or focus. That cap is deliberate: xterm instances sharing a font and theme
+ * share a WebGL texture atlas, and a grid of live terminals would exhaust
+ * Chromium's GL contexts (see `TerminalPanel`'s header comment).
+ */
+export function MissionControlView() {
```
Comment (parsakhaz) [inline frontend/src/components/missionControl/MissionControlView.tsx:216]: Standing: this setter is already behind the effect's cancellation flag (`if (cancelled || ...) return` immediately after the await, with `cancelled` set in the cleanup, and the guard is now also released on effect teardown). That is the guard the rule asks for.

File: frontend/src/components/missionControl/MissionControlView.tsx [hunk 13]
```diff
@@ -0,0 +1,981 @@
...
+    return ids;
+  }, [visibleTiles, focusedPanelId]);
+  // Sorted: the set is what the poll depends on, but `groups` reorders on any
+  // status change, and an order-sensitive key restarted the interval on
+  // ordinary churn.
+  const panelIdsKey = [...panelIds].sort().join(',');
+
+  // Poll snapshots only while this view is actually on screen. A background
+  // poll would keep every agent's emulator warm for nothing.
+  useEffect(() => {
```
Comment (parsakhaz) [inline frontend/src/components/missionControl/MissionControlView.tsx:475]: Standing: this setter is already behind the effect's cancellation flag (`if (cancelled || ...) return` immediately after the await, with `cancelled` set in the cleanup, and the guard is now also released on effect teardown). That is the guard the rule asks for.

File: frontend/src/components/missionControl/MissionControlView.tsx (hunk 12 above)
Comment (parsakhaz) [reply frontend/src/components/missionControl/MissionControlView.tsx:216]: Correcting myself: the previous reply on this thread was posted against the wrong rule, apologies. The answer for no-giant-component is:

Not split in this PR, and I have stopped calling it purely a heuristic. The file is around 900 lines and two rounds of defects came out of it, so a split is the obvious next step. I did not want to attempt it at the end of a long change set, where the regression risk would land on the reviewer rather than on me. It is named in the PR body under what is deliberately left, so the decision is visible rather than buried in a resolved thread.

**Review by parsakhaz — COMMENTED** [review]
## Review pass 1: correctness and security

Three Must-Fix flow-control defects were found. No security defects were found.

1. **Remote terminal acknowledgements bypass designated-viewer arbitration.** `frontend/src/remote/runtime/remoteRuntimeAdapter.ts:187` sends `terminal:ack` without the viewer ID, while visibility registration includes it. `main/src/services/terminalPanelManager.ts:836` credits unidentified acknowledgements without checking the designated viewer, and `main/src/daemon/httpApiServer.ts:785` scopes only visibility calls. With desktop and Remote PWA watching one PTY, the same chunk can be credited twice, defeating the high watermark.

2. **Mission Control acknowledges chunks before xterm consumes them.** `frontend/src/hooks/useMissionControlTerminal.ts:535` credits queued pre-hydration output immediately, and the hydrated path credits immediately after scheduling `terminal.write`. The primary terminal correctly acknowledges from xterm's write callback at `frontend/src/components/panels/TerminalPanel.tsx:1548`. A busy agent can therefore outrun the tile parser without triggering backpressure.

3. **Mission Control's acknowledgement unit differs from main's debit unit.** Main debits `data.length` at `main/src/services/terminalPanelManager.ts:712`, but Mission Control and the Remote PWA credit UTF-8 byte length. CJK or emoji output credits more than was debited and repeatedly clamps the pending counter to zero.

Required fixes: carry and identically scope the remote viewer ID for acknowledgement; acknowledge Mission Control output only after xterm's callback, using bounded batching; use one unit on both sides and add multibyte regression coverage.

Checks: `git diff --check origin/main...HEAD`; `pnpm typecheck`; `pnpm lint`; 78 focused frontend tests; 132 focused main tests. All passed before fixes.

**Comment by parsakhaz** [comment]
## Review, simplify, refactor, and QA complete

Head: `59189e2c`

### Review

The correctness and security review found three Must-Fix defects in terminal flow control. All are fixed in `e2be817a` (`fix(review): preserve terminal flow control`), and the second review passed with zero Must-Fix or Should-Fix findings.

- Remote PWA acknowledgements now carry the registered viewer ID and receive the same authenticated runtime scoping as visibility.
- Mission Control and Remote PWA acknowledge only after xterm consumes output.
- Main, primary terminal, Mission Control, and Remote PWA now use the same UTF-8 byte unit.
- Regression tests cover remote scoping, viewer propagation, delayed batching, disposal flushing, and multibyte output.

GitHub does not allow an author to submit “request changes” on their own PR, so the Must-Fix review was posted as a formal comment review with the same findings.

### Simplify

`5d05670e` (`refactor(simplify): reduce Mission Control state branches`) made two behavior-neutral reductions across the full touched-file review:

- One generation counter now owns stale roster request invalidation instead of two counters that always moved together.
- One derived `canInteract` predicate now owns tile live-render and activation eligibility.

The before and after frontend baseline is identical: 33 files and 342 tests passed, with typecheck and lint clean.

### Refactor

`59189e2c` (`refactor: centralize terminal viewer invariants`) applied only changes with direct test proof:

- Secondary terminals now preserve exact PTY widths above the old 400-column cap. New tests cover 401 and 512 columns.
- Primary TerminalPanel and Mission Control share one tested acknowledgement accumulator.
- The unread `MissionControlSnapshotResult.capturedAt` field was removed from main, shared types, and the E2E mock.

The deep verification finished at 9.8/10 with zero criticals or warnings.

### Tests run

- `pnpm typecheck`: passed
- `pnpm lint`: passed, including Oxlint, ESLint, boundary conformance, and Knip
- Frontend Vitest: 33 files, 344 tests passed on final head
- Focused main Vitest: 6 files, 163 tests passed on final head
- Review-fix focused suites: 12 frontend and 74 main tests passed
- `pnpm test:ci:minimal`: 27 Playwright tests passed
- Focused Mission Control Playwright journey: 1 passed


### UI QA evidence

Synthetic marker: `Mission Control QA`, three mocked agent panes. No production or private data was used.

| State | Evidence |
| --- | --- |
| Populated project grid | Three panes render in the new top-level view with project grouping and density controls.<br>Populated Mission Control grid |
| Status grouping and 4× request | Controls switch grouping and requested density while the responsive cap keeps tiles readable.<br>Mission Control grouped by status |
| Collapsed group | The status group collapses and reports its count.<br>Collapsed Mission Control status group |
| 420px viewport | With the app's compact sidebar, controls wrap inside the viewport and the document has no horizontal overflow.<br>Mission Control at 420 pixels wide |

The uploaded PNGs were downloaded again and matched their source SHA-256 hashes.


### Follow-ups

- Add component-level orchestration tests before splitting the 1,046-line `MissionControlView` into roster, polling, selection, keyboard, and rendering units.
- Add terminal lifecycle tests before introducing a shared desktop, Mission Control, and Remote PWA terminal-viewer boundary.
- Measure and migrate agent roster identity out of panel-state JSON, with backfill and query-plan tests, before optimizing the roster query.
- Decide Cursor's Mission Control grouping and labels, then reuse the canonical agent type with grouping coverage.

### Left for parsa

- All required GitHub checks passed on macOS, Ubuntu, and Windows.
- React Doctor remains advisory at 70/100. It reports one render-purity error for
  updating the tile-model cache ref during `useMemo`, plus five structural
  warnings. No required check failed and no behavior regression was demonstrated,
  so this was not expanded into an unrequested fourth pass.
- Optionally repeat the real-agent hover, focus, typing, and terminal-query checks on this final head. The existing PR manual record covers them before these three commits.
- Merge when satisfied. No merge was performed here.
